pkg/compiler/validators/typing.go

```diff
@@ -68,29 +46,57 @@ func (t trivialChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
 	return upstreamTypeCopy.String() == downstreamTypeCopy.String()
 }
 
-// The void type matches only void
-func (t voidChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
-	return isVoid(upstreamType)
+type noneTypeChecker struct{}
+
+// CastsFrom matches only void
+func (t noneTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
+	return isNoneType(upstreamType)
 }
 
-// For a map type checker, we need to ensure both the key types and value types match.
+type mapTypeChecker struct {
+	literalType *flyte.LiteralType
+}
+
+// CastsFrom checks that the target map type can be cast to the current map type. We need to ensure both the key types
+// and value types match.
 func (t mapTypeChecker) CastsFrom(upstreamType *flyte.LiteralType) bool {
+	// Empty maps should match any collection.
+	if isNoneType(upstreamType.GetMapValueType()) {
+		return true
+	}
+
 	mapLiteralType := upstreamType.GetMapValueType()
 	if mapLiteralType != nil {
 		return getTypeChecker(t.literalType.GetMapValueType()).CastsFrom(mapLiteralType)
 	}
+
 	return false
```

Comment:
Does this make it a little more readable?
```suggestion
	mapLiteralType := upstreamType.GetMapValueType()
	if isNoneType(mapLiteralType) {
		// Empty maps should match any collection.
		return true
	} else if mapLiteralType != nil {
		return getTypeChecker(t.literalType.GetMapValueType()).CastsFrom(mapLiteralType)
	}

	return false
```